fixes #135, #137, and #201 - custom fields choices refactor

lampwins opened the pull request:
<!--
    Thank you for your interest in contributing to Nautobot! Please note
    that our contribution policy recommends that a feature request or bug
    report be opened for approval prior to filing a pull request. This
    helps avoid wasting time and effort on something that we might not
    be able to accept.

    Please indicate the relevant feature request or bug report below.
-->
### Fixes: #135 #137 #201 

As discussed, this breaks out choices into their own model. Changes to choices and cleaning of state CF data now run in the RQ worker under the `custom_fields` queue.

Also changing the name or type of the custom field are not permitted once created.

There are two things I am not happy with and would like some input:
1. change logging is not currently possible because the jobs are queued from the model's save and delete methods and/or signals.
2. The integrity check for choices in the admin panel causes a weird UX flow. If a choice is in use or is the default for the field, it cannot be deleted. The check on the delete method raises a ProtectedError, however, this is being manually handled in the admin `save_formsets` which is only capable of setting user messages. This means the delete fails correctly, but the user is kicked back to the CF list view with a green message saying the choice could not be deleted.

glennmatthews commented on nautobot/extras/models/customfields.py:
```diff
@@ -325,8 +338,87 @@ def validate(self, value):
+                model = ct.model_class()
+                if model.objects.filter(**{f"custom_field_data__{self.field.name}__contains": self.value}).exists():
+                    raise models.ProtectedError(self, "Cannot delete this choice because it is in active use.")
+
```
From a UX perspective, it's a bit odd that we can delete a custom field altogether even if it's in active use, but can't delete a custom field choice if it's in use. Would it make sense or be desirable to allow deletion of choices and have it revert all instances of that choice to either the default value or to the empty value instead?

jathanism replied:
IMO it should also be impossible to delete a custom field if it's in active use and continue to assert the behavior that if you want to nuke all that data at once, you have to do it the hard way.

glennmatthews commented on nautobot/extras/tasks.py:
```diff
@@ -0,0 +1,53 @@
+                new_list = [new_value if e == old_value else e for e in old_list]
+                obj.custom_field_data[field.name] = new_list
+                obj.save()
+
```
A couple of thoughts here:

- Would it make sense to create a `JobResult` for either of these new tasks?
- Should this task log an error or failure of some sort if the `field` is not found or if the `field.type` is not in `(TYPE_SELECT, TYPE_MULTISELECT)`?

lampwins replied:
To the first point, I think that is a more fundamental question we need to address in the core. I see these as internal application jobs and thus exposing JobResults that get mixed with userspace jobs would be confusing to end-users. These jobs can still be monitored in the RQ queue and this would be the same in a celery world. Note that the release check job follows this same pattern today.

glennmatthews commented on nautobot/extras/tasks.py:
```diff
@@ -0,0 +1,53 @@
+            model = ct.model_class()
+            for obj in model.objects.filter(**{f"custom_field_data__{field.name}": old_value}):
+                obj.custom_field_data[field.name] = new_value
+                obj.save()
```
What would be required to let us use `web_request_context` for change logging and webhooks on these tasks? Is it just a matter of making it so that `user` is an optional rather than mandatory parameter to that?

lampwins replied:
Yep, that is pretty much it if we wanted to follow the present pattern. I have also wondered about a couple of packages out there that expose the request object wherever you want, akin to the way Flask works. I assume they work by throwing it in thread-local storage. (like this project which looks like someone's night time pet project https://github.com/schecterdamien/django-request-context).

lampwins replied:
@jathanism @glennmatthews any further thoughts on this? We could do what Glenn suggested above but it would mean all those change log records never have a user associated.

glennmatthews commented on nautobot/extras/tests/test_customfields.py:
```diff
@@ -116,6 +119,40 @@ def test_select_field(self):
+        # Retrieve the stored value
+        site.refresh_from_db()
+        self.assertEqual(site.custom_field_data[cf.name], ["Option A", "Option B"])
+
```
Perhaps some testing around successfully/unsuccessfully deleting `CustomFieldChoice` instances here?

glennmatthews commented on nautobot/extras/signals.py:
```diff
@@ -104,14 +105,14 @@ def handle_cf_removed_obj_types(instance, action, pk_set, **kwargs):
     """
     if action == "post_remove":
-        instance.remove_stale_data(ContentType.objects.filter(pk__in=pk_set))
+        delete_custom_field_data.delay(instance.name, pk_set)
```
Huh. Given that we have `CASCADE` deletion on `CustomField -> CustomFieldChoice`, can we successfully delete a `TYPE_SELECT` `CustomField` that is in use on various models, or do the prohibitions against deleting an in-use `CustomFieldChoice` get in the way? I *think* it should work but it'd be great to see explicit testing around this.

lampwins replied:
Yes, this works because Django's `CASCADE` ignores the subordinate's `delete()` method. I added some tests though.

glennmatthews commented on nautobot/extras/tests/test_customfields.py:
```diff
@@ -758,8 +832,7 @@ def setUpTestData(cls):
             name="cf8",
-            type=CustomFieldTypeChoices.TYPE_URL,
-            choices=["Foo", "Bar", "Baz"],
+            type=CustomFieldTypeChoices.TYPE_SELECT,
```
Are we missing somewhere where the `choices` should be created for this field?

Thetacz commented on nautobot/core/templates/inc/custom_fields_panel.html:
```diff
@@ -15,7 +15,9 @@
-                            {% elif value is not None %}
+                            {% elif field.type == 'multi-select' and value %}
+                                {{ value|join:", " }}
+                            {% elif value is not None and value %}
```
Sorry to step in the review process, isn't it cleaner to just show the intention?
```suggestion
                            {% elif value or value == 0 %}
```

jathanism replied:
`value is not None and value` would never catch `value == 0`. So if the need is to explicitly act upon the integer `0`, then I agree (since that is equivalent to `False`).

Thanks @Thetacz, don't ever apologize for contributing. :)

jathanism commented on nautobot/docs/additional-features/custom-fields.md:
```diff
@@ -39,6 +40,8 @@ Each custom selection field must have at least two choices. These are specified
 
```
"All primary models support custom fields"?

jathanism commented on nautobot/extras/api/views.py:
```diff
@@ -80,6 +80,12 @@ class CustomFieldViewSet(ModelViewSet):
     filterset_class = filters.CustomFieldFilterSet
 
 
+class CustomFieldChoiceViewSet(ModelViewSet):
```
Yes, also appears in browsable API views.

jathanism commented on nautobot/extras/models/customfields.py:
```diff
@@ -325,8 +338,87 @@ def validate(self, value):
+                    )
+
+            if self.type == CustomFieldTypeChoices.TYPE_MULTISELECT:
+                if not set(value).issubset(self.choices.values_list("value", flat=True)):
```
Beautiful.

jathanism commented on nautobot/extras/models/customfields.py:
```diff
@@ -168,6 +164,16 @@ def remove_stale_data(self, content_types):
+            # Check immutable fields
+            database_object = self.__class__.objects.get(pk=self.pk)
+
+            if self.name != database_object.name:
+                raise ValidationError({"name": "Name cannot be changed once created"})
+
+            if self.type != database_object.type:
+                raise ValidationError({"type": "Type cannot be changed once created"})
```
Seems to me we might also make set `editable=False` on the `name` and `type` fields to coincide with this validation?

That would simplify the edit views perhaps?

lampwins replied:
My understanding of editable is it just excludes the field from inclusion in forms and model validation. In the admin pannel we could implement `get_readonly_field()` but I would rather not, and focus on a solution once we move away from the admin pannel.

jathanism replied:
Works for me.

glennmatthews commented on nautobot/extras/models/customfields.py:
```diff
@@ -325,8 +328,97 @@ def validate(self, value):
+        to="extras.CustomField",
+        on_delete=models.CASCADE,
+        related_name="choices",
+        limit_choices_to={"type": CustomFieldTypeChoices.TYPE_SELECT},
```
Doesn't this need to include TYPE_MULTISELECT as well?

lampwins replied:
Hmm, presumably yes, however, it is working now, so I am not sure what this is actually doing then...

lampwins replied:
So according to the docs, this only affects choices in forms and the admin. Because of the UX of the formsets in admin in this case, it doesn't actually matter. I will add it for now so it is not confusing, but we will revisit this anyway with a new UI for custom fields.

glennmatthews commented on nautobot/extras/models/customfields.py:
```diff
@@ -325,8 +328,97 @@ def validate(self, value):
+        super().save(*args, **kwargs)
+
+        if self.value != database_object.value:
+            update_custom_field_choice_data.delay(self.field.pk, database_object.value, self.value)
```
This can throw a `rq.connections.NoRedisConnectionException` if you have an outdated `nautobot_config.py` that's defining `RQ_QUEUES` without `"custom_fields"` queue, as I've discovered. :-) Maybe add some error checks here?

lampwins replied:
Yeah, good catch!

lampwins replied:
I have addressed this with a new check E006

glennmatthews commented on nautobot/core/settings.py:
```diff
@@ -479,4 +479,12 @@
         "SSL": False,
         "DEFAULT_TIMEOUT": 300,
     },
+    "custom_fields": {
```
Need to update `nautobot/core/templates/nautobot_config.py.j2`, `development/nautobot_config.py`, and `nautobot/core/tests/nautobot_config.py` with this new queue as well.

lampwins commented on nautobot/extras/models/customfields.py:
```diff
@@ -154,20 +151,19 @@ class Meta:
+
         # Validate the field's default value (if any)
         if self.default is not None:
             try:
```
So right now it is a two-step process. Create the field, create the choices, and then update the field with the default. I discussed yesterday that we could work around this with a hacky solution that involves moving this validation to the view. I don't like it for obvious reasons, and it would not be consistent with the API.

glennmatthews replied:
Thanks for jogging my memory. :-) I can make it a two-step process in the importer as well since we already do two passes of synchronization anyway for other reasons.

jathanism commented on nautobot/extras/models/customfields.py:
```diff
@@ -325,8 +328,97 @@ def validate(self, value):
         elif self.required:
             raise ValidationError("Required field cannot be empty.")
+
+    def delete(self, *args, **kwargs):
+        """
+        Handle the cleanup of old custom field data when a CustomField is deleted.
+        """
+        content_types = set(self.content_types.values_list("pk", flat=True))
+
+        super().delete(*args, **kwargs)
+
+        delete_custom_field_data.delay(self.name, content_types)
```
Just revisiting the discussion from yesterday: At first glance it seems a little weird to me to overload `delete()` to fire the background task vs. registering a `pre_delete` signal to call it? Since `delete()` isn't called during bulk operations (even tho we have limited use of those) and signals are, just curious about the design tradeoffs here.

The other concern that comes to mind is if there's a need for other internal apps to be aware of when custom field values are deleted. If so, signals might be the better path?

lampwins replied:
So my opinion on signals is they should primarily only be used for cross-app decoupling. Whenever you use signals to implement logic that is entirely internal to a single app (as is the case here with deleting a custom field), I find that to be an anti-pattern because it does nothing more than obscure the code path.

> The other concern that comes to mind is if there's a need for other internal apps to be aware of when custom field values are deleted. If so, signals might be the better path?

Nothing stops another app from registering their own signal for this event on custom fields, and this would actually be the correct usage of a signal IMO.

The only exception to the above that I hold is for `m2m_changed` because it offers abilities that otherwise are hard to implement in other code paths. I think you basically said the same thing yesterday?

Bulk ORM operations is a valid concern, however, in nautobot we explicitly avoid any such usage primarily due to the changelogging implementation. Would love to discuss that, but I see that separate from this thread.

jathanism replied:
Good stuff. Thanks!

jathanism commented on nautobot/extras/models/customfields.py:
```diff
@@ -325,8 +328,97 @@ def validate(self, value):
+        if self.value != database_object.value:
+            update_custom_field_choice_data.delay(self.field.pk, database_object.value, self.value)
+
+    def delete(self, *args, **kwargs):
+        """
+        When a custom field choice is deleted, remove references to in custom field data
+        """
+        if self.field.default:
+            # Cannot delete the choice if it is the default value.
+            if self.field.type == CustomFieldTypeChoices.TYPE_SELECT and self.field.default == self.value:
+                raise models.ProtectedError(
+                    self, "Cannot delete this choice because it is the default value for the field."
+                )
+            elif self.value in self.field.default:
+                raise models.ProtectedError(
+                    self, "Cannot delete this choice because it is one of the default values for the field."
+                )
+
+        if self.field.type == CustomFieldTypeChoices.TYPE_SELECT:
+            # Check if this value is in active use in a select field
+            for ct in self.field.content_types.all():
+                model = ct.model_class()
+                if model.objects.filter(**{f"custom_field_data__{self.field.name}": self.value}).exists():
+                    raise models.ProtectedError(self, "Cannot delete this choice because it is in active use.")
+
+        else:
+            # Check if this value is in active use in a multi-select field
+            for ct in self.field.content_types.all():
+                model = ct.model_class()
+                if model.objects.filter(**{f"custom_field_data__{self.field.name}__contains": self.value}).exists():
+                    raise models.ProtectedError(self, "Cannot delete this choice because it is in active use.")
```
Same concerns here re: `delete()` overload vs. using signals.

lampwins replied:
In this case, we are actually implementing integrity checks that _will_ veto a delete. To me, that calls for a more direct code path. Because of the UX flow here, we have to explicitly call `delete()` within the formset save logic and catch any exceptions. Thus, moving this logic to a signal does nothing but obscure that code flow IMO.

lampwins commented:
@jathanism @glennmatthews In the latest commit I have done two things:

- Refactored all migrations
- Renamed the `custom_field_data` field to `_custom_field_data` to denote it as private which the intention of exposing the `cf` property as the official interface for custom field instance data. This is a larger effort that I want to get into 1.0.0 but not into this PR. For now, I have added a `custom_field_data` property to make this a little easier.

glennmatthews commented on nautobot/extras/tests/test_customfields.py:
```diff
@@ -817,3 +927,130 @@ def test_filter_url(self):
+        cf.save()
+        cf.content_types.set([obj_type])
+
+        # Synchronously process all jobs on the queue in this process
```
👍

glennmatthews commented on nautobot/extras/tests/test_customfields.py:
```diff
@@ -817,3 +927,130 @@ def test_filter_url(self):
+
+        site.refresh_from_db()
+
+        self.assertTrue("cf1" in site.cf)
```
Can we define a default value for `cf1` and confirm that it's set here?

jathanism commented on nautobot/core/checks.py:
```diff
@@ -67,3 +69,10 @@ def check_storage_config_and_backend(app_configs, **kwargs):
+
+@register(Tags.compatibility)
+def check_custom_fields_queue_defined(app_configs, **kwargs):
+    if settings.RQ_QUEUES and not settings.RQ_QUEUES.get("custom_fields"):
```
Seems like a simple change. Could also leverage `Error.hint` to say which one resulted in the error.

jathanism commented on nautobot/extras/filters.py:
```diff
@@ -101,11 +107,51 @@ def __init__(self, *args, **kwargs):
+
+    class Meta:
+        model = CustomFieldChoice
+        fields = ["id", "value", "weight"]
```
@glennmatthews From the [docs](https://django-filter.readthedocs.io/en/stable/ref/filterset.html#declaring-filterable-fields):

> Note that it is not necessary to included declared filters in a `fields` list - doing so will have no effect - and including declarative aliases in a `fields` dict will raise an error.

lampwins replied:
This pattern has fixed implementation in the django ecosystem which is annoying. For instance, either django-filters and/or django-tables take the opposite stance.

jathanism commented on nautobot/extras/filters.py:
```diff
@@ -101,11 +107,51 @@ def __init__(self, *args, **kwargs):
+        method="search",
+        label="Search",
+    )
+    content_types = ContentTypeMultipleChoiceFilter(
```
Yes. But it's also fixing the other extras filtersets (tag, status, web hook) by swapping the inheritance from `CustomFieldFilterSet` to `CustomFieldModelFilterSet`.

jathanism commented on nautobot/extras/api/customfields.py:
```diff
@@ -87,3 +88,11 @@ def _populate_custom_fields(self, instance, custom_fields):
             instance.custom_fields[field.name] = instance.cf.get(field.name)
+
+
+class CustomFieldChoiceSerializer(ValidatedModelSerializer):
```
Looks like this one should go away.

lampwins commented on nautobot/extras/filters.py:
```diff
@@ -79,12 +81,16 @@ def __init__(self, custom_field, *args, **kwargs):
             if custom_field.filter_logic == CustomFieldFilterLogicChoices.FILTER_LOOSE:
                 self.lookup_expr = "icontains"
 
+        if custom_field.type == CustomFieldTypeChoices.TYPE_MULTISELECT:
```
So turns out the answer is no because `lookup_expr` is overridden by the call to `super().__init__()`

glennmatthews commented on development/nautobot_config.py:
```diff
@@ -97,6 +97,14 @@ def is_truthy(arg):
+        "PASSWORD": os.environ["REDIS_PASSWORD"],
+        "SSL": is_truthy(os.environ.get("REDIS_SSL", False)),
+        "DEFAULT_TIMEOUT": 900,
+    },
```
No `webhooks` queue here?

Also I just realized that these new queues also need to be added to `nautobot/docs/configuration/required-settings.md` as well.

glennmatthews commented on nautobot/docs/additional-features/custom-fields.md:
```diff
@@ -20,6 +20,9 @@ Custom fields must be created through the admin UI under Extras > Custom Fields.
 Each custom field must have a name; this should be a simple database-friendly string, e.g. `tps_report`. You may also assign a corresponding human-friendly label (e.g. "TPS report"); the label will be displayed on web forms. A weight is also required: Higher-weight fields will be ordered lower within a form. (The default weight is 100.) If a description is provided, it will appear beneath the fie ...
 
+!!! note
+    The name and type of a custom field cannot be modified once created, so take care in defining the name and type. This helps to reduce the possibility of inconsistent data and enforces the importance of thinking about the network data model when defining a new custom fields.
```
```suggestion
    The name and type of a custom field cannot be modified once created, so take care in defining the name and type. This helps to reduce the possibility of inconsistent data and enforces the importance of thinking about the network data model when defining a new custom field.
```

glennmatthews reviewed (approved):
:shipit:
